src/devices/hid-keyboard.c

```diff
@@ -0,0 +1,188 @@
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "i2c-hid.h"
+#include "hid_api.h"
+#include "spinlock.h"
+#include "utils.h"
+#include "bit_ops.h"
+
+#define MAX_PRESSED_KEYS 6
+
+static const uint8_t keyboard_hid_report_descriptor[] = {
+    0x05, 0x01,     /* Usage Page (Generic Desktop) */
+    0x09, 0x06,     /* Usage (Keyboard) */
+    0xa1, 0x01,     /* Collection (Application) */
+    0x75, 0x01,     /*   Report Size (1) */
+    0x95, 0x08,     /*   Report Count (8) */
+    0x05, 0x07,     /*   Usage Page (Key Codes) */
+    0x19, 0xe0,     /*   Usage Minimum (224) */
+    0x29, 0xe7,     /*   Usage Maximum (231) */
+    0x15, 0x00,     /*   Logical Minimum (0) */
+    0x25, 0x01,     /*   Logical Maximum (1) */
+    0x81, 0x02,     /*   Input (Data, Variable, Absolute) */
+    0x95, 0x01,     /*   Report Count (1) */
+    0x75, 0x08,     /*   Report Size (8) */
+    0x81, 0x01,     /*   Input (Constant) */
+    0x95, 0x05,     /*   Report Count (5) */
+    0x75, 0x01,     /*   Report Size (1) */
+    0x05, 0x08,     /*   Usage Page (LEDs) */
+    0x19, 0x01,     /*   Usage Minimum (1) */
+    0x29, 0x05,     /*   Usage Maximum (5) */
+    0x91, 0x02,     /*   Output (Data, Variable, Absolute) */
+    0x95, 0x01,     /*   Report Count (1) */
+    0x75, 0x03,     /*   Report Size (3) */
+    0x91, 0x01,     /*   Output (Constant) */
+    0x95, 0x06,     /*   Report Count (6) */
+    0x75, 0x08,     /*   Report Size (8) */
+    0x15, 0x00,     /*   Logical Minimum (0) */
+    0x25, 0xff,     /*   Logical Maximum (255) */
+    0x05, 0x07,     /*   Usage Page (Key Codes) */
+    0x19, 0x00,     /*   Usage Minimum (0) */
+    0x29, 0xff,     /*   Usage Maximum (255) */
+    0x81, 0x00,     /*   Input (Data, Array) */
+    0xc0,       /* End Collection */
+};
+
+struct hid_keyboard {
+    hid_dev_t hid_dev;
+    
+    uint32_t pad[1024];
+
+    spinlock_t lock;
+
+    uint8_t input_report[10];
+    uint8_t output_report[3];
+
+    // state
+    uint32_t keys[8];
+    uint32_t leds;
+};
+
+
+static void hid_keyboard_reset(void* dev)
+{
+    hid_keyboard_t* kb = (hid_keyboard_t*)dev;
+    spin_lock(&kb->lock);
+    memset(kb->keys, 0, sizeof(kb->keys));
+    kb->leds = 0;
+    spin_unlock(&kb->lock);
+}
+
+static void hid_keyboard_read_report(void* dev,
+    uint8_t report_type, uint8_t report_id, uint32_t offset, uint8_t *val)
+{
+    (void)report_id;
+    hid_keyboard_t* kb = (hid_keyboard_t*)dev;
+    spin_lock(&kb->lock);
+    switch (report_type) {
+    case REPORT_TYPE_INPUT:
+        if (offset == 0) {
+            kb->input_report[0] = bit_cut(sizeof(kb->input_report), 0, 8);
+            kb->input_report[1] = bit_cut(sizeof(kb->input_report), 8, 8);
+            kb->input_report[2] = bit_cut(kb->keys[7], 0, 8);
+            kb->input_report[3] = 0;
+            uint32_t idx = 0;
+            for (uint32_t code = 0; code < 256; code++) {
```

Comment:
Is there a way to optimize this somehow? I.e. not loop over all buttons on report read, and buffer pressed keys in `hid_keyboard_press()`

Not a huge performance concern, but why not)